api/src/main/java/marquez/db/SearchDao.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package marquez.db;
+
+import java.util.List;
+import marquez.api.models.SearchFilter;
+import marquez.api.models.SearchResult;
+import marquez.api.models.SearchSort;
+import marquez.db.mappers.SearchResultMapper;
+import org.jdbi.v3.sqlobject.config.RegisterRowMapper;
+import org.jdbi.v3.sqlobject.statement.SqlQuery;
+
+/** The DAO for {@link SearchResult}. */
+@RegisterRowMapper(SearchResultMapper.class)
+public interface SearchDao {
+  /**
+   * Returns all datasets and jobs that match the provided query; matching of datasets and jobs is
+   * string based and case-insensitive.
+   *
+   * @param query Query containing pattern to match.
+   * @param filter The filter to apply to the query result.
+   * @param sort The sort to apply to the query result.
+   * @param limit The limit to apply to the query result.
+   * @return A {@link SearchResult} object.
+   */
+  @SqlQuery(
+      "SELECT type, name, updated_at, namespace_name\n"
+          + "FROM (\n"
+          + "  SELECT 'DATASET' AS type, d.name, d.updated_at, d.namespace_name\n"
+          + "    FROM datasets AS d\n"
+          + "   WHERE  d.name ilike '%' || :query || '%'\n"
+          + "   UNION\n"
+          + "  SELECT 'JOB' AS type, j.name, j.updated_at, j.namespace_name\n"
+          + "    FROM jobs AS j\n"
+          + "   WHERE  j.name ilike '%' || :query || '%'\n"
```

Comment:
I think we might need some testing for an empty string here. I believe it will return everything. Maybe also worth mentioning in the API docs.

Reply:
I think we'd want to return no results if the string is empty? I can add a check to immediately return if `q=`.

Reply:
Ok, that works as well.